File: Python/206.Reverse Linked List.py
#把链表逆序
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None
class Solution(object):
    def reverseList(self, head):
        if head==None:  return None
        pre , cur = head , head.next
        pre.next = None
        while cur:
            post = cur.next
            cur.next = pre
            pre = cur
            cur = post
        return pre



l1 = ListNode(1)
l2 = ListNode(2)
l3 = ListNode(6)
l4 = ListNode(3)
l5 = ListNode(5)
l6 = ListNode(4)
l7 = ListNode(7)
l1.next = l2
l2.next = l3
l3.next = l4
l4.next = l5
l5.next = l6
l6.next = l7

s = Solution()
p = s.reverseList(l1)
while p:
    print(p.val)
    p = p.next

File: C++/Linked List/206.Reverse Linked List.cpp
/*
把链表逆序。
*/

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* reverseList(ListNode* head) {
		if(!head)	return NULL;
        ListNode nHead(0);
        nHead.next = head;
        ListNode *pre = head, *cur = head->next;
        while(cur)
        {
        	pre->next = cur->next;
        	cur->next = nHead.next;
        	nHead.next = cur;
        	cur = pre->next;
        }
        return nHead.next;
    }
};
int main()
{
	Solution s;
	ListNode head(1);
	ListNode t1(4);
	ListNode t2(3);
	ListNode t3(2);
	ListNode t4(5);
	ListNode t5(2);
	head.next = &t1;
	t1.next = &t2;
	t2.next = &t3;
	t3.next = &t4;
	t4.next = &t5;
 	ListNode* h = s.reverseList(&head);
 	while(h)
 	{
 		cout<<h->val<<endl;
 		h = h->next;
 	}
}

